comparison/scripts/nbconverted/0.get-cytominer-tool-differences.py

```diff
@@ -0,0 +1,544 @@
+# Cytominer is an R based image-based profiling tool.
+# In this repo, we reprocess the data with [pycytominer](https://github.com/cytomining/pycytominer).
+# As the name connotes, pycytominer is a python based image-based profiling tool.
+# 
+# We include all processing scripts and present the pycytominer profiles in this open source repository.
+# The repository represents a unified bioinformatics pipeline applied to all Cell Painting Drug Repurposing Profiles. In this notebook, we compare the resulting output data between the processing pipelines for the two tools: Cytominer and pycytominer.
+# We output several metrics comparing the two approaches
+# 
+# ## Metrics
+# 
+# In all cases, we calculate the element-wise absolute value difference between pycytominer and cytominer profiles.
+# 
+# 1. Mean, median, and sum of element-wise differencs
+# 2. Per feature mean, median, and sum of element-wise differences
+# 3. Feature selection procedure differences per feature (level 4b only)
+# 
+# In addition, we confirm alignment of the following metadata columns:
+# 
+# * Well
+# * Broad Sample Name
+# * Plate
+# 
+# Other metadata columns are not expected to be aligned.
+# For example, we have [updated MOA and Target information](https://github.com/broadinstitute/lincs-cell-painting/issues/11) in the pycytominer version.
+# 
+# ## Data Levels
+# 
+# Image-based profiling results in the following output data levels.
+# We do not compare all data levels in this notebook.
+# 
+# | Data | Level | Comparison |
+# | :---- | :---- | :-------- |
+# | Images | Level 1 | NA |
+# | SQLite File (single cell profiles ) | Level 2 | NA |
+# | Aggregated Profiles with Well Information (metadata) | Level 3 | Yes  |
+# | Normalized Aggregated Profiles with Metadata | Level 4a | Yes | 
+# | Normalized and Feature Selected Aggregated Profiles with Metadata | Level 4b | Yes |
+# | Perturbation Profiles created Summarizing Replicates  | Level 5 | No |
+
+# In[1]:
+
+
+get_ipython().run_line_magic('load_ext', 'nb_black')
+
+
+# In[2]:
+
+
+import os
+import pathlib
+import numpy as np
+import pandas as pd
+
+from util import build_file_dictionary, load_data, generate_output_filenames
+
+
+# In[3]:
+
+
+# Pycytominer plates are saved with 5 floating point decimals
+round_decimals = 5
+
+# Create the output directory
+output_dir = "results"
+os.makedirs(output_dir, exist_ok=True)
+
+
+# In[4]:
+
+
+def get_metrics(pycyto_df, cyto_df, features):
+    # Align features
+    pycyto_df = pycyto_df.reindex(features, axis="columns")
+    cyto_df = cyto_df.reindex(features, axis="columns")
+
+    # Assess difference
+    abs_diff = pycyto_df.subtract(cyto_df).abs()
+    mean_diff = abs_diff.mean()
+    median_diff = abs_diff.median()
+    sum_diff = abs_diff.sum()
+
+    complete_mean_diff = mean_diff.replace([np.inf, -np.inf], np.nan).dropna().mean()
```

Comment:
(This question comes from my lack of knowledge of what are the range of values of each feature) Why are there `np.inf` values in the mean/median/sum values? Do `pycto_df` or `cyto_df` contain `np.inf` values?

Reply:
Thanks @niranjchandrasekaran - indeed, you've touched upon a pretty important difference between python and R-based processing. Solving this problem is beyond scope of this PR, but it does likely explain many of the small floating point differences between cytominer and pycytominer profiles.

I outlined the issue in cytomining/pycytominer#79 - although I was not sure where it belongs since it permeates so many different codebases.

Reply:
> Why are there np.inf values in the mean/median/sum values? Do pycto_df or cyto_df contain np.inf values?

Neither of them contain `np.inf` values. Some features are entirely NA in cytominer, but this is recoded as 0 in pycytominer. The `np.inf` happens in this specific column after subtraction and summarization.

so, `pycyto_df.subtract(cyto_df).abs().mean()` will result in `np.inf` for features with that case described above.